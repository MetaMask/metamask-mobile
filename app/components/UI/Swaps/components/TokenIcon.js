import React, { useCallback, useState } from 'react';
import PropTypes from 'prop-types';
import { StyleSheet, View } from 'react-native';

import RemoteImage from '../../../Base/RemoteImage';
import Text from '../../../Base/Text';
import { useTheme } from '../../../../util/theme';
import imageIcons from '../../../../images/image-icons';

/* eslint-disable import/no-commonjs */
const ethLogo = require('../../../../images/eth-logo-new.png');
/* eslint-enable import/no-commonjs */

const REGULAR_SIZE = 24;
const REGULAR_RADIUS = 12;
const MEDIUM_SIZE = 36;
const MEDIUM_RADIUS = 18;
const BIG_SIZE = 50;
const BIG_RADIUS = 25;
const BIGGEST_SIZE = 70;
const BIGGEST_RADIUS = 35;

const createStyles = (colors) =>
  StyleSheet.create({
    icon: {
      width: REGULAR_SIZE,
      height: REGULAR_SIZE,
      borderRadius: REGULAR_RADIUS,
    },
    iconMedium: {
      width: MEDIUM_SIZE,
      height: MEDIUM_SIZE,
      borderRadius: MEDIUM_RADIUS,
    },
    iconBig: {
      width: BIG_SIZE,
      height: BIG_SIZE,
      borderRadius: BIG_RADIUS,
    },
    iconBiggest: {
      width: BIGGEST_SIZE,
      height: BIGGEST_SIZE,
      borderRadius: BIGGEST_RADIUS,
    },
    emptyIcon: {
      backgroundColor: colors.background.alternative,
      alignItems: 'center',
      justifyContent: 'center',
    },
    tokenSymbol: {
      fontSize: 16,
      textAlign: 'center',
      textAlignVertical: 'center',
      color: colors.text.default,
    },
    tokenSymbolMedium: {
      fontSize: 22,
      color: colors.text.default,
    },
    tokenSymbolBig: {
      fontSize: 26,
      color: colors.text.default,
    },
  });

const EmptyIcon = ({ medium, big, biggest, style, ...props }) => {
  const { colors } = useTheme();
  const styles = createStyles(colors);

  return (
    <View
      style={[
        styles.icon,
        medium && styles.iconMedium,
        big && styles.iconBig,
        biggest && styles.iconBiggest,
        styles.emptyIcon,
        style,
      ]}
      {...props}
    />
  );
};

EmptyIcon.propTypes = {
  medium: PropTypes.bool,
  big: PropTypes.bool,
  biggest: PropTypes.bool,
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
};

function TokenIcon({ symbol, icon, medium, big, biggest, style }) {
  const [showFallback, setShowFallback] = useState(false);
  const { colors } = useTheme();
  const styles = createStyles(colors);

  const getSource = useCallback(() => {
    if (symbol === 'ETH') {
      return ethLogo;
    }

    if (Object.keys(imageIcons).includes(symbol)) {
      return imageIcons[symbol];
    }

    if (icon) {
      return { uri: icon };
    }

    return null;
  }, [symbol, icon]);
  const source = getSource();

  if (source && !showFallback) {
    return (
      <RemoteImage
        fadeIn
        source={getSource()}
        onError={() => setShowFallback(true)}
        style={[
          styles.icon,
          medium && styles.iconMedium,
          big && styles.iconBig,
          biggest && styles.iconBiggest,
          style,
        ]}
      />
    );
  }

  if (symbol) {
    return (
      <EmptyIcon medium={medium} big={big} biggest={biggest} style={style}>
        <Text
          style={[
            styles.tokenSymbol,
            medium && styles.tokenSymbolMedium,
            (big || biggest) && styles.tokenSymbolBig,
            biggest && styles.tokenSymbolBiggest,
          ]}
        >
          {symbol[0].toUpperCase()}
        </Text>
      </EmptyIcon>
    );
  }

  return <EmptyIcon medium={medium} style={style} />;
}

TokenIcon.propTypes = {
  symbol: PropTypes.string,
  icon: PropTypes.string,
  medium: PropTypes.bool,
  big: PropTypes.bool,
  biggest: PropTypes.bool,
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
};

export default TokenIcon;
