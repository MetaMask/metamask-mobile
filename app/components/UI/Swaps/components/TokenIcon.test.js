import React from 'react';
import { shallow } from 'enzyme';
import TokenIcon from './TokenIcon';

describe('TokenIcon component', () => {
  it('should Render correctly', () => {
    const empty = shallow(<TokenIcon />);
    expect(empty).toMatchSnapshot();
    const eth = shallow(<TokenIcon symbol="ETH" />);
    expect(eth).toMatchSnapshot();
    const symbol = shallow(<TokenIcon symbol="cDAI" />);
    expect(symbol).toMatchSnapshot();
    const icon = shallow(
      <TokenIcon
        symbol="DAI"
        icon="https://cloudflare-ipfs.com/ipfs/QmNYVMm3iC7HEoxfvxsZbRoapdjDHj9EREFac4BPeVphSJ"
      />,
    );
    expect(icon).toMatchSnapshot();
    const emptyMedium = shallow(<TokenIcon medium />);
    expect(emptyMedium).toMatchSnapshot();
    const ethMedium = shallow(<TokenIcon medium symbol="ETH" />);
    expect(ethMedium).toMatchSnapshot();
    const symbolMedium = shallow(<TokenIcon medium symbol="cDAI" />);
    expect(symbolMedium).toMatchSnapshot();
    const iconMedium = shallow(
      <TokenIcon
        medium
        symbol="DAI"
        icon="https://cloudflare-ipfs.com/ipfs/QmNYVMm3iC7HEoxfvxsZbRoapdjDHj9EREFac4BPeVphSJ"
      />,
    );
    expect(iconMedium).toMatchSnapshot();
  });
});
