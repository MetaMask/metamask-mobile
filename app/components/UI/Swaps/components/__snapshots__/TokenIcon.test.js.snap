// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenIcon component should Render correctly 1`] = `<EmptyIcon />`;

exports[`TokenIcon component should Render correctly 2`] = `
<RemoteImage
  fadeIn={true}
  onError={[Function]}
  source={
    Object {
      "testUri": "../../../app/images/eth-logo-new.png",
    }
  }
  style={
    Array [
      Object {
        "borderRadius": 12,
        "height": 24,
        "width": 24,
      },
      undefined,
      undefined,
      undefined,
      undefined,
    ]
  }
/>
`;

exports[`TokenIcon component should Render correctly 3`] = `
<EmptyIcon>
  <Text
    style={
      Array [
        Object {
          "color": "#24272A",
          "fontSize": 16,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    C
  </Text>
</EmptyIcon>
`;

exports[`TokenIcon component should Render correctly 4`] = `
<RemoteImage
  fadeIn={true}
  onError={[Function]}
  source={
    Object {
      "uri": "https://cloudflare-ipfs.com/ipfs/QmNYVMm3iC7HEoxfvxsZbRoapdjDHj9EREFac4BPeVphSJ",
    }
  }
  style={
    Array [
      Object {
        "borderRadius": 12,
        "height": 24,
        "width": 24,
      },
      undefined,
      undefined,
      undefined,
      undefined,
    ]
  }
/>
`;

exports[`TokenIcon component should Render correctly 5`] = `
<EmptyIcon
  medium={true}
/>
`;

exports[`TokenIcon component should Render correctly 6`] = `
<RemoteImage
  fadeIn={true}
  onError={[Function]}
  source={
    Object {
      "testUri": "../../../app/images/eth-logo-new.png",
    }
  }
  style={
    Array [
      Object {
        "borderRadius": 12,
        "height": 24,
        "width": 24,
      },
      Object {
        "borderRadius": 18,
        "height": 36,
        "width": 36,
      },
      undefined,
      undefined,
      undefined,
    ]
  }
/>
`;

exports[`TokenIcon component should Render correctly 7`] = `
<EmptyIcon
  medium={true}
>
  <Text
    style={
      Array [
        Object {
          "color": "#24272A",
          "fontSize": 16,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        Object {
          "color": "#24272A",
          "fontSize": 22,
        },
        undefined,
        undefined,
      ]
    }
  >
    C
  </Text>
</EmptyIcon>
`;

exports[`TokenIcon component should Render correctly 8`] = `
<RemoteImage
  fadeIn={true}
  onError={[Function]}
  source={
    Object {
      "uri": "https://cloudflare-ipfs.com/ipfs/QmNYVMm3iC7HEoxfvxsZbRoapdjDHj9EREFac4BPeVphSJ",
    }
  }
  style={
    Array [
      Object {
        "borderRadius": 12,
        "height": 24,
        "width": 24,
      },
      Object {
        "borderRadius": 18,
        "height": 36,
        "width": 36,
      },
      undefined,
      undefined,
      undefined,
    ]
  }
/>
`;
